improved hashtag detection with Twitter's strategy

adrianmc opened the pull request:
As per #247, I have extracted Twitter's hashtag detection strategy and incorporated it into our hashtag plugin. Now, I want to go on the record to state that I'm not sure if this the correct way forward, but at least it's a step forward to get the discussion going.

I welcome more discussion before merging this PR.

In any case, @mxstbr should be happy to know that german words are now detected properly. Feel free to try it with `#österreich`.

mxstbr commented:
Yessssss instant merge!!

nikgraf commented on draft-js-hashtag-plugin/src/utils/index.js:
```diff
@@ -0,0 +1,36 @@
+  return tags;
+}
+
+export function extractHashtags(text) {
```
Is this function used? Can we remove it?

adrianmc replied:
I used this for testing initially. It's not necessarily for the plugin, I'll take it out.

nikgraf commented on draft-js-hashtag-plugin/CHANGELOG.md:
```diff
@@ -5,6 +5,12 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## To Be Released
 
+## 1.0.1
```
If we follow http://semver.org/ I would consider this a MINOR change and suggest 1.1.0. What are your thoughts?

adrianmc replied:
Yes, that makes sense. I'll make the change.

adrianmc commented:
@nikgraf I've made the changes. Feel free to merge if you think it's ready.

nikgraf commented:
@adrianmc are there some tests from the Twitter guys we can re-use? Would be great to have them in here as well.

adrianmc commented:
@nikgraf good question, it seems the JS directory only has the following tests:

```js
var text = "\uD801\uDC00 #hashtag \uD801\uDC00 #hashtag";
var extracted = twttr.txt.extractHashtagsWithIndices(text);
deepEqual(extracted, [{hashtag:"hashtag", indices:[3, 11]}, {hashtag:"hashtag", indices:[15, 23]}], "Hashtag w/ Supplementary character, UTF-16 indices");
twttr.txt.modifyIndicesFromUTF16ToUnicode(text, extracted);
deepEqual(extracted, [{hashtag:"hashtag", indices:[2, 10]}, {hashtag:"hashtag", indices:[13, 21]}], "Hashtag w/ Supplementary character, Unicode indices");
twttr.txt.modifyIndicesFromUnicodeToUTF16(text, extracted);
deepEqual(extracted, [{hashtag:"hashtag", indices:[3, 11]}, {hashtag:"hashtag", indices:[15, 23]}], "Hashtag w/ Supplementary character, UTF-16 indices");
```

But digger further into the overall repo, I was able to find the following MASSIVE list of tests:

```yaml
  hashtags:
    - description: "Extract an all-alpha hashtag"
      text: "a #hashtag here"
      expected: ["hashtag"]

    - description: "Extract a letter-then-number hashtag"
      text: "this is #hashtag1"
      expected: ["hashtag1"]

    - description: "Extract a number-then-letter hashtag"
      text: "#1hashtag is this"
      expected: ["1hashtag"]

    - description: "DO NOT Extract an all-numeric hashtag"
      text: "On the #16 bus"
      expected: []

    - description: "DO NOT Extract a single numeric hashtag"
      text: "#0"
      expected: []

    - description: "Extract hashtag after bracket"
      text: "(#hashtag1 )#hashtag2 [#hashtag3 ]#hashtag4 ’#hashtag5’#hashtag6"
      expected: ["hashtag1", "hashtag2", "hashtag3", "hashtag4", "hashtag5", "hashtag6"]

    - description: "Extract a hashtag containing ñ"
      text: "I'll write more tests #mañana"
      expected: ["mañana"]

    - description: "Extract a hashtag containing é"
      text: "Working remotely #café"
      expected: ["café"]

    - description: "Extract a hashtag containing ü"
      text: "Getting my Oktoberfest on #münchen"
      expected: ["münchen"]

    - description: "DO NOT Extract a hashtag containing Japanese"
      text: "this is not valid: # 会議中 ハッシュ"
      expected: []

    - description: "Extract a hashtag in Korean"
      text: "What is #트위터 anyway?"
      expected: ["트위터"]

    - description: "Extract a half-width Hangul hashtag"
      text: "Just random half-width Hangul #ﾣﾦﾰ"
      expected: ["ﾣﾦﾰ"]

    - description: "Extract a hashtag in Russian"
      text: "What is #ашок anyway?"
      expected: ["ашок"]

    - description: "Extract a starting katakana hashtag"
      text: "#カタカナ is a hashtag"
      expected: ["カタカナ"]

    - description: "Extract a starting hiragana hashtag"
      text: "#ひらがな FTW!"
      expected: ["ひらがな"]

    - description: "Extract a starting kanji hashtag"
      text: "#漢字 is the future"
      expected: ["漢字"]

    - description: "Extract a trailing katakana hashtag"
      text: "Hashtag #カタカナ"
      expected: ["カタカナ"]

    - description: "Extract a trailing hiragana hashtag"
      text: "Japanese hashtags #ひらがな"
      expected: ["ひらがな"]

    - description: "Extract a trailing kanji hashtag"
      text: "Study time #漢字"
      expected: ["漢字"]

    - description: "Extract a central katakana hashtag"
      text: "See my #カタカナ hashtag?"
      expected: ["カタカナ"]

    - description: "Extract a central hiragana hashtag"
      text: "Study #ひらがな for fun and profit"
      expected: ["ひらがな"]

    - description: "Extract a central kanji hashtag"
      text: "Some say #漢字 is the past. what do they know?"
      expected: ["漢字"]

    - description: "Extract a Kanji/Katakana mixed hashtag"
      text: "日本語ハッシュタグテスト #日本語ハッシュタグ"
      expected: ["日本語ハッシュタグ"]

    - description: "Extract a hashtag after a punctuation"
      text: "日本語ハッシュテスト。#日本語ハッシュタグ"
      expected: ["日本語ハッシュタグ"]

    - description: "DO NOT include a punctuation in a hashtag"
      text: "#日本語ハッシュタグ。"
      expected: ["日本語ハッシュタグ"]

    - description: "Extract a full-width Alnum hashtag"
      text: "全角英数字ハッシュタグ ＃ｈａｓｈｔａｇ１２３"
      expected: ["ｈａｓｈｔａｇ１２３"]

    - description: "DO NOT extract a hashtag without a preceding space"
      text: "日本語ハッシュタグ#日本語ハッシュタグ"
      expected: []

    - description: "Hashtag with chouon"
      text: "長音ハッシュタグ。#サッカー"
      expected: ["サッカー"]

    - description: "Hashtag with half-width chouon"
      text: "長音ハッシュタグ。#ｻｯｶｰ"
      expected: ["ｻｯｶｰ"]

    - description: "Hashtag with half-widh voiced sounds marks"
      text: "#ﾊｯｼｭﾀｸﾞ #ﾊﾟﾋﾟﾌﾟﾍﾟﾎﾟ"
      expected: ["ﾊｯｼｭﾀｸﾞ", "ﾊﾟﾋﾟﾌﾟﾍﾟﾎﾟ"]

    - description: "Hashtag with half-width # after full-width ！"
      text: "できましたよー！#日本語ハッシュタグ。"
      expected: ["日本語ハッシュタグ"]

    - description: "Hashtag with full-width ＃ after full-width ！"
      text: "できましたよー！＃日本語ハッシュタグ。"
      expected: ["日本語ハッシュタグ"]

    - description: "Hashtag with ideographic iteration mark"
      text: "#云々 #学問のすゝめ #いすゞ #各〻 #各〃"
      expected: ["云々", "学問のすゝめ", "いすゞ", "各〻", "各〃"]

    - description: "Extract hashtag with fullwidth tilde"
      text: "#メ～テレ ハッシュタグ内で～が認識されず"
      expected: ["メ～テレ"]

    - description: "Extract hashtag with wave dash"
      text: "#メ〜テレ ハッシュタグ内で～が認識されず"
      expected: ["メ〜テレ"]

    - description: "Hashtags with ş (U+015F)"
      text: "Here’s a test tweet for you: #Ateş #qrşt #ştu #ş"
      expected: ["Ateş", "qrşt", "ştu", "ş"]

    - description: "Hashtags with İ (U+0130) and ı (U+0131)"
      text: "Here’s a test tweet for you: #İn #ın"
      expected: ["İn", "ın"]

    - description: "Hashtag before punctuations"
      text: "#hashtag: #hashtag; #hashtag, #hashtag. #hashtag! #hashtag?"
      expected: ["hashtag", "hashtag", "hashtag", "hashtag", "hashtag", "hashtag"]

    - description: "Hashtag after punctuations"
      text: ":#hashtag ;#hashtag ,#hashtag .#hashtag !#hashtag ?#hashtag"
      expected: ["hashtag", "hashtag", "hashtag", "hashtag", "hashtag", "hashtag"]

    - description: "Hashtag before newline"
      text: "#hashtag\ntest\n#hashtag2\ntest\n#hashtag3\n"
      expected: ["hashtag", "hashtag2", "hashtag3"]

    - description: "DO NOT extract hashtag when # is followed by URL"
      text: "#http://twitter.com #https://twitter.com"
      expected: []

    - description: "DO NOT extract hashtag if it's a part of URL"
      text: "http://twitter.com/#hashtag twitter.com/#hashtag"
      expected: []

    - description: "Extract hashtags with Latin extended characters"
      text: "#Azərbaycanca #mûǁae #Čeština #Ċaoiṁín"
      expected: ["Azərbaycanca", "mûǁae", "Čeština", "Ċaoiṁín"]

    - description: "Extract Arabic hashtags"
      text: "#سیاست #ایران #السياسة #السياح #لغات  #اتمی  #کنفرانس #العربية #الجزيرة #فارسی"
      expected: ["سیاست", "ایران", "السياسة", "السياح", "لغات", "اتمی", "کنفرانس", "العربية", "الجزيرة", "فارسی"]

    - description: "Extract Arabic hashtags with underscore"
      text: "#برنامه_نویسی  #رییس_جمهور  #رئيس_الوزراء, #ثبت_نام. #لس_آنجلس"
      expected: ["برنامه_نویسی", "رییس_جمهور", "رئيس_الوزراء", "ثبت_نام", "لس_آنجلس"]

    - description: "Extract Hebrew hashtags"
      text: "#עַל־יְדֵי #וכו׳ #מ״כ"
      expected: ["עַל־יְדֵי", "וכו׳", "מ״כ"]

    - description: "Extract Thai hashtags"
      text: "#ผู้เริ่ม #การเมือง #รายละเอียด #นักท่องเที่ยว #ของขวัญ #สนามบิน #เดินทาง #ประธาน"
      expected: ["ผู้เริ่ม", "การเมือง", "รายละเอียด", "นักท่องเที่ยว", "ของขวัญ", "สนามบิน", "เดินทาง", "ประธาน"]

    - description: "Extract Arabic hashtags with Zero-Width Non-Joiner"
      text: "#أي‌بي‌إم #می‌خواهم"
      expected: ["أي‌بي‌إم", "می‌خواهم"]

    - description: "Extract Amharic hashtag"
      text: "የአላህ መልእክተኛ ሰለላሁ ዓለይሂ ወሰለም #ኢትዮሙስሊምስ"
      expected: ["ኢትዮሙስሊምስ"]

    - description: "Extract Sinhala hashtag with Zero-Width Joiner (U+200D)"
      text: "#ශ්‍රීලංකා"
      expected: ["ශ්‍රීලංකා"]

    - description: "Extract Arabic and Persian hashtags with numbers"
      text: "#۳۴۵هشتگ #هشتگ۶۷۸ #ســـلام_عليكم_٤٠٦"
      expected: ["۳۴۵هشتگ","هشتگ۶۷۸","ســـلام_عليكم_٤٠٦"]

    - description: "Extract Hindi hashtags"
      text: "#महात्मा #महात्मा_१२३४ #१२३४ गांधी"
      expected: ["महात्मा","महात्मा_१२३४"]

    - description: "Extract Indic script hashtags"
      text: "#বাংলা #ગુજરાતી #ಕನ್ನಡ #മലയാളം #ଓଡ଼ିଆ #ਪੰਜਾਬੀ #සිංහල #தமிழ் #తెలుగు"
      expected: ["বাংলা","ગુજરાતી","ಕನ್ನಡ","മലയാളം","ଓଡ଼ିଆ","ਪੰਜਾਬੀ","සිංහල","தமிழ்","తెలుగు"]

    - description: "Extract Tibetan hashtags"
      text: "#བོད་སྐད་ #བོད་སྐད"
      expected: ["བོད་སྐད་","བོད་སྐད"]

    - description: "Extract Khmer, Burmese, Laotian hashtags"
      text: "#មហាត្មះគន្ធី #မြင့်မြတ်သော #ຊີວະສາດ"
      expected: ["មហាត្មះគន្ធី","မြင့်မြတ်သော","ຊີວະສາດ"]

    - description: "Extract Greek hashtag"
      text: "#Μαχάτμα_Γκάντι ήταν Ινδός πολιτικός"
      expected: ["Μαχάτμα_Γκάντι"]

    - description: "Extract Armenian and Georgian hashtags"
      text: "#Մահաթմա #მაჰათმა"
      expected: ["Մահաթմա","მაჰათმა"]

    - description: "Extract hashtag with middle dot"
      text: "#il·lusió"
      expected: ["il·lusió"]

    - description: "DO NOT extract hashtags without a letter"
      text: "#_ #1_2 #122 #〃"
      expected: []

  hashtags_with_indices:
    - description: "Extract a hastag at the start"
      text: "#hashtag here"
      expected:
        - hashtag: "hashtag"
          indices: [0, 8]

    - description: "Extract a hastag at the end"
      text: "test a #hashtag"
      expected:
        - hashtag: "hashtag"
          indices: [7, 15]

    - description: "Extract a hastag in the middle"
      text: "test a #hashtag in a string"
      expected:
        - hashtag: "hashtag"
          indices: [7, 15]

    - description: "Extract only a valid hashtag"
      text: "#123 a #hashtag in a string"
      expected:
        - hashtag: "hashtag"
          indices: [7, 15]

    - description: "Extract a hashtag in a string of multi-byte characters"
      text: "会議中 #hashtag 会議中"
      expected:
        - hashtag: "hashtag"
          indices: [4, 12]

    - description: "Extract multiple valid hashtags"
      text: "One #two three #four"
      expected:
        - hashtag: "two"
          indices: [4, 8]
        - hashtag: "four"
          indices: [15, 20]

    - description: "Extract a non-latin hashtag"
      text: "Hashtags in #русский!"
      expected:
        - hashtag: "русский"
          indices: [12, 20]

    - description: "Extract multiple non-latin hashtags"
      text: "Hashtags in #中文, #日本語, #한국말, and #русский! Try it out!"
      expected:
        - hashtag: "中文"
          indices: [12, 15]
        - hashtag: "日本語"
          indices: [17, 21]
        - hashtag: "한국말"
          indices: [23, 27]
        - hashtag: "русский"
          indices: [33, 41]
```

That's a hell of a lot of tests though. I don't know if we should incorporate all of them?

nikgraf commented:
@adrianmc yeah, let's do it. I think it's better if we have them as well. In case Twitter's & our code diverges at some point or gets improved we can easily verify.
